Pruned mode sync peer selection

neonknight64 opened the pull request:
## Description
The sync peer selection mechanism was changed to allow the node listening state to only select sync peers with the entire range of data required by the current node. The pruning horizon and the current nodes tip height is used to determine if the network node has all the required chain data.

## Motivation and Context
These changes allow block syncing to function correctly when some nodes are running in archival mode. It will enable a lagging node to only select sync peers with all the chain data required for syncing by considering the reported tip height and pruning horizon of network nodes.

## How Has This Been Tested?
 Two tests were added for the node listening state to test sync peer selection and sync mode selection.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [x] New Tests
* [ ] Documentation

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
* [x] I'm merging against the `development` branch.
* [x] I ran `cargo-fmt --all` before pushing.
* [x] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.

sdbondi commented on base_layer/core/src/base_node/states/listening.rs:
```diff
@@ -174,3 +179,96 @@ fn determine_sync_mode(
+        let sync_peers = select_sync_peers(local_tip_height, &best_network_metadata, &peer_metadata_list);
+        assert_eq!(sync_peers.len(), 0);
+
+        let node_id1 = NodeId::try_from(&[1; NODE_ID_ARRAY_SIZE][..]).unwrap();
```
Why not just use a random NodeId?

neonknight64 replied:
Change applied, made use a random NodeId.

sdbondi reviewed (commented):
Looks good - just a question about why you exported the NodeId const from comms

Commit:
- These changes allow block syncing to function correctly when some nodes are running in archival mode. It will enable a lagging node to only select sync peers with all the chain data required for syncing by considering the reported tip height and pruning horizon of network nodes.
- The sync peer selection mechanism was changed to allow the node listening state to only select sync peers with the entire range of data required by the current node. The pruning horizon and the current nodes tip height is used to determine if the network node has all the required chain data.
- Two tests were added for the node listening state to test sync peer selection and sync mode selection.

sdbondi reviewed (approved):
LGTM
